smart_open/s3.py

```diff
@@ -807,7 +833,8 @@ def _upload_next_part(self):
         logger.debug("upload of part #%i finished" % part_num)
 
         self._total_parts += 1
-        self._buf = io.BytesIO()
+        self._buf.close()
```

Comment:
Is the temp file removal safe even if something goes wrong? We don't want to pollute the disk on errors.

Reply:
I refactored my solution. The user now passes in their own buffer (if they want to, otherwise nothing changes).

Reply:
Ah, you shifted the problem on to the user. Clever :)